File: huggingface_scraper/ipfs_kit_lib/ipfs_cluster_follow.js
import { execSync, exec } from 'child_process';
import os from 'os';
import path, { parse } from 'path';
import fs from 'fs';
import * as install_ipfs from './install_ipfs.js';

export class IPFSClusterFollow {
    constructor(resources, meta) {
        // Default values
        this.config = {};
        this.role = 'leecher'; // default role
        this.clusterName = '';

        // Set from meta if provided
        if (meta.config) {
            this.config = meta.config;
        }

        if (meta.role && ['master', 'worker', 'leecher'].includes(meta.role)) {
            this.role = meta.role;
        } else if (meta.role) {
            throw new Error("role is not either master, worker, leecher");
        }

        if (meta.ipfs_path) {
            this.ipfsPath = meta.ipfs_path;
        }

        if (meta.cluster_name) {
            this.clusterName = meta.cluster_name;
        }
    }

    async ipfsFollowStart(clusterName = this.clusterName) {
        let results = { systemctl: false, bash: false };
        let detectResults
        if (os.userInfo().uid === 0) {
            try {
                // Attempt to start the ipfs-cluster-follow service
                let systemctlStart = execSync("systemctl start ipfs-cluster-follow").toString();
                results.systemctl = systemctlStart;
                // Check if the service is running
                detectResults = execSync("ps -ef | grep ipfs-cluster-follow | grep -v grep").toString().trim();
                        
                if (detectResults.length === 0) {
                    const homeDir = os.homedir();
                    const followDir = path.join(homeDir, ".ipfs-cluster-follow", clusterName);
                    // Check for and remove the api-socket file if it exists
                    const apiSocketPath = path.join(followDir, "api-socket");
                    if (fs.existsSync(apiSocketPath)) {
                        fs.unlinkSync(apiSocketPath);
                        results.bash = true;
                    }
                    // Attempt to start the ipfs-cluster-follow service
                    let systemctlStart = execSync("systemctl start ipfs-cluster-follow").toString();
                    results.systemctl = systemctlStart;
                    // Check if the service is running
                    detectResults = execSync("ps -ef | grep ipfs-cluster-follow | grep -v grep").toString().trim();
                }
            }
            catch (error) {
                console.log(`Error starting ipfs-cluster-follow: ${error.message}`);
                results.systemctl = `Error: ${error.message}`;
            }
        } 
        if (os.userInfo().uid != 0 || results.systemctl.includes('Error') || detectResults.length === 0) {
            try {            
                // Attempt to run the ipfs-cluster-follow command
                const commandRun = `/usr/local/bin/ipfs-cluster-follow ${clusterName} run`;
                const process = exec(commandRun, (error, stdout, stderr) => {
                    if (error) {
                        console.error(`Error running ipfs-cluster-follow: ${error.message}`);
                        if (error.message.includes("command not found")){
                            let InstallIPFS = new install_ipfs.InstallIPFS();
                            let installResults = InstallIPFS.installIPFSClusterFollow();
                            let configResults = InstallIPFS.configIPFSClusterFollow();
                            console.log(`Install IPFS Cluster Follow results: ${installResults}`);
                            console.log(`Configure IPFS Cluster Follow results: ${configResults}`);
                            console.log(`Attempting to run ipfs-cluster-follow again`);
                            exec(commandRun, (error, stdout, stderr) => {
                                if (error) {
                                    console.error(`Error running ipfs-cluster-follow: ${error.message}`);
                                }
                                console.log(`ipfs-cluster-follow output: ${stdout}`);
                            });
                        }
                        return;
                    }
                    console.log(`ipfs-cluster-follow output: ${stdout}`);
                });
            }
            catch (error) {
                console.log(`Error running ipfs-cluster-follow: ${error.message}`);
                console.error(`Error in ipfsFollowStart: ${error.message}`);
            }
        } 
        return results;
    }

    async ipfsFollowStop() {
        let results = { systemctl: '', bash: '', 'api-socket': '' };
        let clusterName = this.clusterName;
        
        if (os.userInfo().uid === 0) {
            try {
                // Attempt to stop the ipfs-cluster-follow service
                const systemctlStop = execSync("systemctl stop ipfs-cluster-follow").toString();
                results.systemctl = systemctlStop;
            } catch (error) {
                console.log(`Error stopping ipfs-cluster-follow: ${error.message}`)
                results.systemctl = `Error: ${error.message}`;
            }
        }
        if (os.userInfo().uid != 0 || results.systemctl.includes('Error')){
            try {
                // Forcefully kill the ipfs-cluster-follow process if it's still running
                const listCommand = "ps -ef | grep ipfs-cluster-follow | grep -v grep | wc -l";
                const listResults = execSync(listCommand).toString().trim();
                if (parseInt(listResults) > 0) {              
                    const killCommand = "ps -ef | grep ipfs-cluster-follow | grep -v grep | awk '{print $2}' | xargs kill -9";
                    const killResults = execSync(killCommand).toString();
                    results.bash = killResults;
                }
                else{
                    results.bash = "No ipfs-cluster-follow process found to kill";
                }
            } catch (error) {
                console.log(`Error killing ipfs-cluster-follow: ${error.message}`)
                results.bash = `Error: ${error.message}`;
            }
        }

        try {
            // Remove the api-socket file
            const apiSocketPath = path.join(os.homedir(), ".ipfs-cluster-follow", clusterName, "api-socket");
            if (fs.existsSync(apiSocketPath)) {
                fs.unlinkSync(apiSocketPath);
                results['api-socket'] = 'Removed api-socket';
            } else {
                //console.log("api-socket not found, deleting not necessary");
                results['api-socket'] = 'api-socket not found, deleting not necessary';
            }
        } catch (error) {
            console.error(`Error removing api-socket: ${error.message}`);
            results['api-socket'] = `Error removing api-socket: ${error.message}`;
        }

        return results;
    }

    async ipfsFollowList(clusterName = this.clusterName) {
        let command = `ipfs-cluster-follow ${clusterName} list`;

        try {
            let results = execSync(command, { encoding: 'utf8' }).trim();

            if (results.length > 0) {
                let resultsArray = results.split("\n");
                let resultsDict = {};

                resultsArray.forEach((result) => {
                    // Replace multiple spaces with a single space
                    result = result.replace(/\s\s+/g, ' ');
                    let parts = result.split(" ");

                    // Assuming the first element is the value and the second is the key
                    if (parts.length >= 2) {
                        let key = parts[1];
                        let value = parts[0];
                        resultsDict[key] = value;
                    }
                });

                return resultsDict;
            } else {
                return false;
            }
        } catch (error) {
            console.error(`Error executing ipfs-cluster-follow list: ${error.message}`);
            return false;
        }
    }


    ipfsFollowInfo(kwargs = {}) {
        let clusterName = this.clusterName;
        let resultsDict = {};

        try {
            const command = `ipfs-cluster-follow ${clusterName} info`;
            let results = execSync(command, { encoding: 'utf8' }).trim().split("\n");

            if (results.length > 0) {
                resultsDict = {
                    clusterName: clusterName,
                    configFolder: results[2].split(": ")[1],
                    configSource: results[3].split(": ")[1],
                    clusterPeerOnline: results[4].split(": ")[1],
                    ipfsPeerOnline: results[5].split(": ")[1],
                };
            }
        } catch (error) {
            console.error(`Error executing ipfs-cluster-follow info: ${error.message}`);
            if (error.message.includes("500")){
                let InstallIPFS = new install_ipfs.InstallIPFS(undefined, { role: 'leecher', cluster_name: clusterName, ipfs_path: this.ipfsPath });
                let installResults = InstallIPFS.installIPFSClusterFollow();
                let configResults = InstallIPFS.configIPFSClusterFollow();
                console.log(`Install IPFS Cluster Follow results: ${installResults}`);
                console.log(`Configure IPFS Cluster Follow results: ${configResults}`);
                console.log(`Attempting to run ipfs-cluster-follow info again`);
                execSync(command, { encoding: 'utf8' });
            }
        }

        return resultsDict;
    }

    async ipfs_follow_run(kwargs = {}) {
        let cluster_name = this.cluster_name;
        if ('cluster_name' in kwargs) {
            cluster_name = kwargs['cluster_name'];
        }

        const command = `ipfs-cluster-follow ${cluster_name} run`;
        let results = execSync(command).toString();
        results = results.split("\n");
        return results;
    }

    async test_ipfs_cluster_follow() {
        let detect;
        try {
            detect = execSync('which ipfs-cluster-follow').toString();
        } catch (error) {
            detect = '';
        }
        return detect.length > 0;
    }

}

async function main(){

    const meta = {
        cluster_name: "test"
    };

    const thisIpfsClusterFollow = new IpfsClusterFollow(meta);
    const results = await thisIpfsClusterFollow.test_ipfs_cluster_follow();
    console.log(results);
}

File: ipfs_transformers/ipfs_kit_lib/ipfs_cluster_follow.py
import sys
import os
import subprocess
import tempfile
import json

class ipfs_cluster_follow:
    def __init__(self, resources, meta=None):
        if meta is not None:
            if "config" in meta:
                if meta['config'] is not None:
                    self.config = meta['config']
            if "role" in meta:
                if meta['role'] is not None:
                    self.role = meta['role']
                    if self.role not in  ["master","worker","leecher"]:
                        raise Exception("role is not either master, worker, leecher")
                    else:
                        self.role = "leecher"
            
            if "cluster_name" in meta:
                if meta['cluster_name'] is not None:
                    self.cluster_name = meta['cluster_name']

            if self.role == "leecher" or self.role == "worker" or self.role == "master":
                pass

    def ipfs_follow_start(self, **kwargs):
        if "cluster_name" in list(self.__dict__.keys()):
            cluster_name = self.cluster_name
        if "cluster_name" in kwargs:
            cluster_name = kwargs['cluster_name']
        try:
            if os.getuid() == 0:
                command1 = "systemctl start ipfs-cluster-follow"
                results1 = subprocess.check_output(command1, shell=True)
                results1 = results1.decode()
            else:
                results1 = "You need to be root to run this command"
        except Exception as e:
            results = str(e)
        finally:
            pass
        detect = "ps -ef | grep ipfs-cluster-follow | grep -v grep | awk '{print $2}'"
        detect_results = subprocess.check_output(detect, shell=True)
        detect_results = detect_results.decode()
        results2 = False

        if len(detect_results) == 0:
            homedir = os.path.expanduser("~")
            ls_file = os.listdir(homedir + "/.ipfs-cluster-follow/" + cluster_name)      
            if "api-socket" in ls_file:
                rm_command = "rm ~/.ipfs-cluster-follow/" + cluster_name + "/api-socket"
                rm_results = subprocess.check_output(rm_command, shell=True)
                rm_results = rm_results.decode()
                results2 = True   
            try:
                command2 = "/usr/local/bin/ipfs-cluster-follow " + cluster_name + " run"
                results2 = subprocess.Popen(command2, shell=True, stdout=subprocess.PIPE)
            except Exception as e:
                results = str(e)
            finally:
                pass

        results = {
            "systemctl": results1,
            "bash": results2
        }
        return results

    def ipfs_follow_stop(self, **kwargs):
        if "cluster_name" in list(self.__dict__.keys()):
            cluster_name = self.cluster_name
        if "cluster_name" in kwargs:
            cluster_name = kwargs['cluster_name']
        try:
            if os.getuid() == 0:
                command1 = "systemctl stop ipfs-cluster-follow"
                results1 = subprocess.check_output(command1, shell=True)
                results1 = results1.decode()
            else:
                results1 = "You need to be root to run this command"
        except Exception as e:
            results1 = str(e)
        finally:
            pass
        try:
            command2 = "ps -ef | grep ipfs-cluster-follow | grep -v grep | awk '{print $2}' | xargs kill -9"
            results2 = subprocess.check_output(command2, shell=True)
            results2 = results2.decode()
        except Exception as e:
            results2 = str(e)
        finally:
            pass

        try:
            command3 = "rm ~/.ipfs-cluster-follow/" + cluster_name + "/api-socket"
            results3 = subprocess.check_output(command3, shell=True)
            results3 = results3.decode()
        except Exception as e:
            results3 = str(e)
        finally:
            pass

        results = {
            "systemctl": results1,
            "bash": results2,
            "api-socket": results3
        }
        return results  

#    def ipfs_follow_run(self, **kwargs):
#        if "cluster_name" in list(self.keys()):
#            cluster_name = self.cluster_name
#        if "cluster_name" in kwargs:
#            cluster_name = kwargs['cluster_name']
#
#        command = "ipfs cluster-follow " + cluster_name + " run"
#        results = subprocess.check_output(command, shell=True)
#        results = results.decode()
#        return results

    def ipfs_follow_list(self, **kwargs):
        if "cluster_name" in list(self.__dict__.keys()):
            cluster_name = self.cluster_name
        if "cluster_name" in kwargs:
            cluster_name = kwargs['cluster_name']

        command = "ipfs-cluster-follow " + cluster_name + " list"
        results = subprocess.check_output(command, shell=True)
        results = results.decode()
        results_dict = {}
        if len(results) > 0:
            results = results.split("\n")
            for i in range(len(results)):

                while "  " in results[i]:
                    results[i] = results[i].replace("  ", " ")

                results[i] = results[i].split(" ")
                if len(results[i]) >= 2:
                    results[i] = {
                        results[i][1]: results[i][0]
                    }

            for i in range(len(results)):
                if type(results[i]) == dict:
                    key = list(results[i].keys())[0]
                    value = results[i][key]
                    results_dict[key] = value
                    
            return results_dict
        else:
            return False

    def ipfs_follow_info(self, **kwargs):
        results_dict = {}
        if "cluster_name" in list(self.__dict__.keys()):
            cluster_name = self.cluster_name
        if "cluster_name" in list(kwargs.keys()):
            cluster_name = kwargs['cluster_name']
        try:
            command = "ipfs-cluster-follow " + cluster_name + " info"
            results = subprocess.check_output(command, shell=True)
            results = results.decode()
            results = results.split("\n")
            if len(results) > 0:
                results_dict = {
                    "cluster_name": cluster_name,
                    "config_folder": results[2].split(": ")[1],
                    "config_source": results[3].split(": ")[1],
                    "cluster_peer_online": results[4].split(": ")[1],
                    "ipfs_peer_online": results[5].split(": ")[1],
                }

        except Exception as e:
            results = str(e)
        finally:
            pass
        
        return results_dict
        
    
    def ipfs_follow_run(self, **kwargs):
        if "cluster_name" in list(self.keys()):
            cluster_name = self.cluster_name
        if "cluster_name" in kwargs:
            cluster_name = kwargs['cluster_name']

        command = "ipfs-cluster-follow "+ cluster_name +" run"
        results = subprocess.check_output(command, shell=True)
        results = results.decode()
        results = results.split("\n")
        return results


    def test_ipfs_cluster_follow(self):
        detect = subprocess.check_output("which ipfs-cluster-follow", shell=True)
        detect = detect.decode()
        if len(detect) > 0:
            return True
        else:
            return False
        pass

if __name__ == "__main__":
    meta = {
        "cluster_name": "test"
    }
    this_ipfs_cluster_follow = ipfs_cluster_follow(meta)
    results = this_ipfs_cluster_follow.test_ipfs_cluster_follow()
    print(results)
    pass
